### PR TITLE
Hook into EVM execution using tracing-like events

### PR DESCRIPTION
**Pull request by nanocryk**
Alternative to #18 . Add a `tracing` feature the gasometer, runtime and main crate that allow to listen to events emitted from important parts of the EVM execution.

### CODE REVIEW

File: gasometer/src/lib.rs [hunk 1]
```diff
@@ -1,10 +1,26 @@
 //! EVM gasometer.
 
 #![deny(warnings)]
-#![forbid(unsafe_code, unused_variables, unused_imports)]
```
Comment (sorpaas) [inline gasometer/src/lib.rs:4]: Can you add back `unused_improts` warning here?

File: gasometer/src/lib.rs [hunk 2]
```diff
@@ -27,6 +43,14 @@ macro_rules! try_or_fail {
 	)
 }
 
+#[derive(Debug, Copy, Clone)]
+pub struct Snapshot {
+    pub gas_limit: u64,
```
Comment (sorpaas) [inline gasometer/src/lib.rs:48]: Indentation

File: src/tracing.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
+//! Allows to listen to runtime events.
+
+use crate::{Context};
```
Comment (sorpaas) [inline src/tracing.rs:3]:
```suggestion
use crate::Context;
```

**Review by sorpaas — APPROVED** [review]
LGTM, but please fix those styling issues.

File: gasometer/src/lib.rs (hunk 1 above)
Comment (nanocryk) [reply gasometer/src/lib.rs:4]: It prevented to use `environmental` macro :/ Tried to allow it locally for the macro call but Rust complained.

File: gasometer/Cargo.toml [hunk 4]
```diff
@@ -12,11 +12,16 @@ edition = "2018"
 primitive-types = { version = "0.9", default-features = false }
 evm-core = { version = "0.26", path = "../core", default-features = false }
 evm-runtime = { version = "0.26", path = "../runtime", default-features = false }
+environmental = { version = "1.1.2", default-features = false, optional = true}
```
Comment (sorpaas) [inline gasometer/Cargo.toml:15]:
```suggestion
environmental = { version = "1.1.2", default-features = false, optional = true }
```

File: gasometer/src/lib.rs [hunk 5]
```diff
@@ -27,6 +43,14 @@ macro_rules! try_or_fail {
...
+    pub memory_gas: u64,
```
Comment (sorpaas) [inline gasometer/src/lib.rs:49]: Indentation still isn't right here, can you fix?

File: Cargo.toml [hunk 6]
```diff
@@ -19,6 +19,7 @@ primitive-types = { version = "0.9", default-features = false, features = ["rlp"
 serde = { version = "1.0", default-features = false, features = ["derive"], optional = true }
 codec = { package = "parity-scale-codec", version = "2.0", default-features = false, features = ["derive"], optional = true }
 ethereum = { version = "0.7", default-features = false }
+environmental = { version = "1.1.2", default-features = false, optional = true}
```
Comment (sorpaas) [inline Cargo.toml:22]:
```suggestion
environmental = { version = "1.1.2", default-features = false, optional = true }
```

**Pull request merged**